### ISSUE TITLE
Index preview broken in Symphony 2.6

### ISSUE THREAD

**Issue by animaux** [symphonists/image_index_preview#20]
In the entries list the file path is rendered:
    http:/root/image/1/0/93/0//img/pathfilename.png

though it obviously should be:
    http:/root/image/1/0/93/0/img/path/filename.png

Preview on publish page works fine though. I guess this should be easy to fix, but I’m lost in if-else clauses …

**Comment by nitriques** [comment]
So the problem is the missing / ?

**Comment by animaux** [comment]
Rather displaced —> //img

**Comment by nitriques** [comment]
Whats the url if you [remove those lines](https://github.com/symphonists/image_index_preview/blob/master/assets/image_index_preview.publish.js#L20-L37)

**Comment by animaux** [comment]
That results in no <img> at all.

**Comment by nitriques** [comment]
Can you replace [those lines](https://github.com/symphonists/image_index_preview/blob/master/assets/image_index_preview.publish.js#L33-L36) with:

```js
// add preview
$('<img />', {
	src: root + '/image/1/' + size + '/' + (!!external ? external + '/' : '') + file
}).prependTo(link);
```

**Comment by nitriques** [comment]
No better: remove the extra / at the end of workspace here ? https://github.com/symphonists/image_index_preview/blob/master/assets/image_index_preview.publish.js#L54

**Comment by animaux** [comment]
Thanks, but neither did work :(

before

    http://server.de/image/1/0/93/0//img/uploads_57-1.jpg

after replacing `// add preview …`

    http://server.de/image/1/0/93//img/uploads_57-1.jpg

Reverting that change and emoving the extra / at the end of workspace results in no change.

**Comment by animaux** [comment]
Like I said before: The preview on publish page works fine though. I guess that doesn’t help …

**Comment by nitriques** [comment]
Damn. Sorry but I do not have this extension installed anywhere so it's kind of hard to spot!

**Comment by animaux** [comment]
Still, many thanks!

Will have a fresh look tomorrow and see if I can sort out what all the if/else-clauses are for :)

**Comment by nitriques** [comment]
Is the path in the html ok ? Look for the href attribute of the link and the data-path attribute...

**Comment by nitriques** [comment]
> if/else-clauses

Mainly differences between Symphony's version...

**Comment by nitriques** [comment]
A hackful fix would be to 

```js
(root + '/image/1/' + size + '/' + external + '/' + file).replace('//', '/')
```

:smile:

**Comment by animaux** [comment]
Oh, I see … I wonder what has changed in comparison to 2.4/2.5 that makes the URL wrong now. 

I’m also beginning to suspect it might have to do with using a regular upload field now. In the past I have used the image_upload field, which has it’s own preview, if I remember correctly.

**Comment by animaux** [comment]
Haha :)

**Comment by animaux** [comment]
`href` and `data-path` seem to look fine:

    <a href="http://server.de/workspace/img/uploads/_57-1.jpg" data-path="/workspace/img/uploads"><img src="http://server.de/image/1/0/93/0//img/uploads_57-1.jpg"></a>

**Comment by animaux** [comment]
@nitriques What works for me in `Symphony 2.6beta2` is replacing [line 48] with:

    file = path.replace(root, '').replace('/workspace/','') + '/' + filename;

Not sure if this breaks things for versions `> 2.6` :(

**Comment by nitriques** [comment]
Ok so L48 will be like L54 then.

> Not sure if this breaks things for versions > 2.6 :(

Me neither. Do you have an old site where you could test ?

**Comment by animaux** [comment]
Hmn, I have used image_upload with most sites. Will check.

**Comment by animaux** [comment]
Ok, I tried this on a `Symphony 2.3.5` installation and it worked fine! :)

**Comment by animaux** [comment]
I dare to send a pull request :)

https://github.com/symphonists/image_index_preview/pull/21

**Comment by nilshoerrmann** [comment]
To be sure that this works across the different upload fields, you should make sure that `filename` doesn't include a slash already.

**Comment by nitriques** [comment]
> you should make sure that filename doesn't include a slash already.

I never saw this from any of the upload fields I've used... Did you ?

**Comment by nilshoerrmann** [comment]
I haven't checked but we shouldn't trust the extensions. This extension breaks every now and then because of "wrong paths" and inconsistencies between upload fields sound like a logical reason to me.

**Comment by nitriques** [comment]
I agree, but it's either this or hacks like this:

```js
(root + '/image/1/' + size + '/' + external + '/' + file).replace('//', '/')
```

Which one to you prefer ?

I do not know why, but mine never broke... https://github.com/Solutions-Nitriques/image_preview/blob/master/assets/image_preview.js  :smile:

**Comment by animaux** [comment]
Hmn, does it do exactly the same and is still working in 2.6? :)

**Comment by nitriques** [comment]
> does it do exactly the same and is still working in 2.6? :)

From what I've seen yes! It should also work with Association Drawer and you can add a setting field to customize the css for it.

**Comment by animaux** [comment]
Will have to check it out. It’s hard to keep track of all the nice extensions :)

**Comment by nitriques** [comment]
Yes it is!!

**Comment by nitriques** [comment]
@animaux Did you had the chance to test on older installs ?

**Comment by animaux** [comment]
https://github.com/symphonists/image_index_preview/issues/20#issuecomment-70109720 ;)

**Comment by nitriques** [comment]
@animaux Oups! Made the comment in the PR :)

**Comment by twiro** [comment]
I just downloaded the current 2.4.1 version of this extension, installed it in a fresh 2.6.1 project and it doesn't work with either the basic "Upload Field" nor with "Unique Upload Field" or "Reflected Upload Field" - still no image previews on index pages.

I then reverted back to [the first fix](https://github.com/symphonists/image_index_preview/commit/7a7f6946250c40c499863e9ef0b70d6bb0c56118) @animaux proposed and everything seemed to work - preview images did show up on index pages.

I decided to take a deeper look and noticed that neither the current nor the previous fix actually do a solid job - depending on wether the uploaded images lie in a subfolder or directly in the workspace-root both fixes fail in getting the path right.

So I wonder - does the current version 2.4.1 of this extension actually work for anyone? If not I'd like to know if my proposal works for you.

Here are my findings from debugging the index preview and the fix that works for me:


### Fix A
_(As introcuded in https://github.com/symphonists/image_index_preview/pull/21 and currently implemented in Version 2.4.1)_

    file = path.replace(root,'').replace('/workspace/','/') + filename;

- transforms **`/workspace/path`** into **`/path`** + **`file.jpg`** = **`/pathfile.jpg`** :x:
- transforms **`/workspace`** into **`/workspace`** + **`file.jpg`** = **`/workspacefile.jpg`** :x:


### Fix B
_(As introduced in https://github.com/symphonists/image_index_preview/pull/17 and removed in https://github.com/symphonists/image_index_preview/pull/19)_

    file = path.replace(root,'').replace('/workspace/','') + '/' + filename;

- transforms **`/workspace/path`** into **`path`** + **`/`** + **`file.jpg`** = **`path/file.jpg`** :white_check_mark:
- transforms **`/workspace`** into **`/workspace`** + **`/`** + **`file.jpg`** = **`/workspace/file.jpg`** :x:


### Fix C
_(PR is on it's way)_

    file = path.replace(root,'').replace('/workspace','') + '/' + filename;

- transforms **`/workspace/path`** into **`/path`** + **`/`** + **`file.jpg`** = **`/path/file.jpg`** :grey_question:
- transforms **`/workspace`** into  **`/`** + **`file.jpg`** = **`/file.jpg`** :grey_question:

This version at least gives us a unique behaviour for both use cases ("workspace" and "subfolder"), but introduces an unwanted slash at the beginning of `file` that results in a double slash (**`//file.jpg`**) before the filename in the final html-output. Doesn't actually seem to be a real issue (at least in firefox), but could easily be avoided by cutting off the first character after transforming the string:

    file = (path.replace(root,'').replace('/workspace','') + '/' + filename).substring(1);

- transforms **`/workspace/path`** into **`/path`** + **`/`** + **`file.jpg`** = **`path/file.jpg`** :white_check_mark:
- transforms **`/workspace`** into  **`/`** + **`file.jpg`** = **`file.jpg`** :white_check_mark:


### Tests & Compatibility

I tested this solution with the basic **Upload Field**, **Unique Upload Field** and **Reflected Upload Field** in Symphony Versions **2.3.6**, **2.5.0** and **2.6.1** and everything looks good.

I also did a quick check with **S3 Upload Field** in Symphony **2.5.0** - looks good to.

**Comment by nitriques** [comment]
Thanks @twiro !

**Comment by twiro** [comment]
Did anyone have the chance to test this?
I integrated my fix in a few symphony installations (Including 2.6.2) on different servers and everything works as expected.
Would be great to see this merged and released as Version 1.4.2!

**Comment by nitriques** [comment]
Should be fixed since I've release 1.4.2. Thanks y'all !
